src/SFML/Audio/Music.cpp

```diff
@@ -39,80 +40,89 @@
 
 namespace sf
 {
+////////////////////////////////////////////////////////////
+struct Music::Impl
+{
+    InputSoundFile            file;     //!< The streamed music file
+    std::vector<std::int16_t> samples;  //!< Temporary buffer of samples
+    std::recursive_mutex      mutex;    //!< Mutex protecting the data
+    Span<std::uint64_t>       loopSpan; //!< Loop Range Specifier
+
+    explicit Impl(InputSoundFile&& theFile) :
+    file(std::move(theFile)),
+
+    // Resize the internal buffer so that it can contain 1 second of audio samples
+    samples(file.getSampleRate() * file.getChannelCount()),
+
+    // Compute the music positions
+    loopSpan{0u, file.getSampleCount()}
+    {
+    }
+};
+
+
 ////////////////////////////////////////////////////////////
 Music::~Music()
 {
     // We must stop before destroying the file
-    stop();
+    if (m_impl != nullptr)
+    {
+        stop();
+    }
 }
 
 
 ////////////////////////////////////////////////////////////
-bool Music::openFromFile(const std::filesystem::path& filename)
-{
-    // First stop the music if it was already running
-    stop();
-
-    // Open the underlying sound file
-    m_file = sf::InputSoundFile::openFromFile(filename);
-    if (!m_file)
-        return false;
+Music::Music(Music&&) noexcept = default;
 
-    // Perform common initializations
-    initialize();
 
-    return true;
-}
+////////////////////////////////////////////////////////////
+Music& Music::operator=(Music&&) noexcept = default;
 
 
 ////////////////////////////////////////////////////////////
-bool Music::openFromMemory(const void* data, std::size_t sizeInBytes)
+std::optional<Music> Music::tryOpenFromFile(std::optional<InputSoundFile>&& optFile)
 {
-    // First stop the music if it was already running
-    stop();
+    if (!optFile.has_value())
+        return std::nullopt;
 
-    // Open the underlying sound file
-    m_file = sf::InputSoundFile::openFromMemory(data, sizeInBytes);
-    if (!m_file)
-        return false;
+    // TODO: apply RVO here via passkey idiom
+    return Music(std::move(*optFile));
+}
 
-    // Perform common initializations
-    initialize();
 
-    return true;
+////////////////////////////////////////////////////////////
+std::optional<Music> Music::openFromFile(const std::filesystem::path& filename)
+{
+    return tryOpenFromFile(sf::InputSoundFile::openFromFile(filename));
```

Comment:
```suggestion
    return tryOpenFromFile(InputSoundFile::openFromFile(filename));
```
Similar comment applies elsewhere in the file.

Reply:
Will change

Reply:
I think you missed this